patroni/validator.py

```diff
@@ -0,0 +1,339 @@
+        elif isinstance(key, Optional):
+            yield key.name
+        elif isinstance(key, Or):
+            if any([i in self.data for i in key.args]):
+                for i in key.args:
+                    if i in self.data:
+                        yield i
+            else:
+                for i in key.args:
+                    yield i
+
+
+def _get_type_name(python_type):
+    if python_type == str:
+        return "a string"
+    elif python_type == int:
+        return "an integer"
+    elif python_type == float:
+        return "a number"
+    elif python_type == bool:
+        return "a boolean"
+    elif python_type == list:
+        return "an array"
+    elif python_type == dict:
+        return "a dictionary"
+    return python_type.__name__
+
+
+def assert_(condition, message="Wrong value"):
+    assert condition, message
+
+
+userattributes = {"username": "", Optional("password"): ""}
+available_dcs = [m.split(".")[-1] for m in dcs_modules()]
+
+schema = Schema({
+  "name": str,
+  "scope": str,
+  "restapi": {
+    "listen": validate_host_port_listen,
+    "connect_address": validate_connect_address
+  },
+  Optional("bootstrap"): {
+    "dcs": {
+        Optional("ttl"): int,
+        Optional("loop_wait"): int,
+        Optional("retry_timeout"): int,
+        Optional("maximum_lag_on_failover"): int
+        },
+    "pg_hba": [str],
+    "initdb": [Or(str, dict)]
+  },
+  Or(*available_dcs): Case({
+      "consul": {
+          Or("host", "url"): Case({
+              "host": str,
+              "url": str})
+          },
+      "etcd": {
+          Or("host", "hosts", "srv", "url", "proxy"): Case({
+              "host": str,
+              "hosts": Or(comma_separated_host_port, [validate_host_port]),
+              "srv": str,
+              "url": str,
+              "proxy": str})
+         },
+      "exhibitor": {
+          "hosts": str,
+          "port": lambda i: assert_(int(i) <= 65535),
+          Optional("pool_interval"): int
+          },
+      "zookeeper": {
+          "hosts": Or(comma_separated_host_port, [validate_host_port]),
+          },
+      "kubernetes": {
+          Optional("namespace"): str,
+          Optional("labels"): {},
+          Optional("scope_label"): str,
+          Optional("role_label"): str,
+          Optional("use_endpoints"): bool,
+          Optional("pod_ip"): Or(is_ipv4_address, is_ipv6_address),
+          Optional("ports"): [{"name": str, "port": int}],
+          },
+      }),
+  "postgresql": {
+    "listen": validate_host_port_listen,
+    "connect_address": validate_connect_address,
+    "authentication": {
+      "replication": userattributes,
+      "superuser": userattributes,
+      "rewind":  userattributes
+    },
+    "data_dir": validate_data_dir,
+    "bin_dir": validate_bin_dir,
+    "parameters": {
+      Optional("unix_socket_directories"): lambda s: assert_(all([isinstance(s, string_types), len(s)]))
+    },
+    Optional("pg_hba"): [str],
```

Comment:
Look, I slightly changed the test:
```diff
--- a/tests/test_validator.py
+++ b/tests/test_validator.py
@@ -59,7 +59,7 @@ config = {
         "parameters": {
             "unix_socket_directories": "."
         },
-        "pg_hba": ["string"],
+        "pg_hba": [u'string'],
         "pg_ident": ["string"],
         "pg_ctl_timeout": 1000,
         "use_pg_rewind": False
```

and such change breaks validator. As I already mentioned we can get the unicode string instead of str when parsing JSON object from DCS.
The following change fixes the issue just for pg_hba case:
```diff
--- a/patroni/validator.py
+++ b/patroni/validator.py
@@ -321,7 +321,7 @@ schema = Schema({
     "parameters": {
       Optional("unix_socket_directories"): lambda s: assert_(all([isinstance(s, string_types), len(s)]))
     },
-    Optional("pg_hba"): [str],
+    Optional("pg_hba"): [string_types],
     Optional("pg_ident"): [str],
     Optional("pg_ctl_timeout"): int,
     Optional("use_pg_rewind"): bool
```

We should do the same for all remaining cases.